src/Databases/DatabaseOnDisk.cpp

```diff
@@ -213,8 +218,47 @@ void DatabaseOnDisk::commitCreateTable(const ASTCreateQuery & query, const Stora
         Poco::File(table_metadata_tmp_path).remove();
         throw;
     }
+
+    try
+    {
+        /// If the table was detached permanently we will have a file with
+        /// .sql_detached suffix, which is not needed anymore since we attached the table back
+        auto table_metadata_file_detached = Poco::File(table_metadata_path + detached_suffix);
+        if (table_metadata_file_detached.exists())
+            table_metadata_file_detached.remove();
```

Comment:
I'm not sure if it's a good idea to silently remove metadata file of permanently detached table. It leaves garbage in `data/` or `store/` directory if permanently detached table has some data on disk. For `Atomic` databases we can use `DatabaseCatalog::enqueueDroppedTableCleanup(...)` to drop it correctly. But maybe just throw an exception if permanently detached table with the same name already exists?

Reply:
Actually right now i did it this way that the ATTACH table becomes normal again and .sql file is recreated back, so `.sql_detached` is not needed anymore (recreating SQL file is a bit safer than renaming it back, because creating happen in stages to prevent races). 

I guess what you are talking is smth like 'ATTACH EPHEMERAL' - i.e. if the table was detached permanently, the attach doesn't return the table to a normal state, it just gets loaded one-time (but will not be loaded with server start).

Reply:
I was thinking about that `ATTACH EPHEMERAL` (it's quite a logical extension of `DETACH PERMANENTLY`), but it seems it may introduce some extra states of the tables which will be harder to follow and will need to break some of the existing invariants (like 'attached table always have own SQL file').

Let's keep it simpler for now and if needed we can add that.

Reply:
```
create database db engine=Atomic;
use db;

-- creates test.sql and store/xxx/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/
create table test (n int) engine=MergeTree order by n; 

-- writes data into store/xxx/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/
insert into test select * from numbers(1000000);

-- renames test.sql to test.sql_detached
detach table test permanently; 

-- creates test.sql with new UUID and store/yyy/yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyy/
-- removes test.sql_detached, but not store/xxx/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/
create table test (n int) engine=MergeTree order by n;
-- now data from store/xxx/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/ can be removed only manually
```

Reply:
![footgun](https://camo.githubusercontent.com/8be9954f1dfeb4944df29d9a9dae5ea3776004650842a07979d8aba00ae00f8b/68747470733a2f2f7374617469632e627573696e657373696e73696465722e636f6d2f696d6167652f3463666536633461343965326165313036373034303030302d3430302f696d6167652e6a7067)

:) 

Isn't it possible already with normal detach?

Reply:
It's not... 
```
Received exception from server (version 20.12.1):
Code: 521. DB::Exception: Received from localhost:9000. DB::Exception: Cannot rename /var/lib/clickhouse/store/78e/78ed63b2-169a-444d-88c7-b77b60f7bfbc/test.sql.tmp to /var/lib/clickhouse/store/78e/78ed63b2-169a-444d-88c7-b77b60f7bfbc/test.sql because the second path already exists, errno: 17, strerror: File exists. 
```

Reply:
I've added some extra checks, see 
https://github.com/ClickHouse/ClickHouse/pull/17642/commits/d9a710db774abd3c0b8b6c104829235866408b21
(will add some tests in a while)

I think that at that moment when table was already attached and has new metadata file (with .SQL extension) it's too late to do something smart with .sql_detached and the best we can do is just remove that file.